A package blog post, 2018-11-11-colocr (#329)
ropensci/roweb2: content/blog/2018-11-11-colocr.md

---
slug: "colocr"
title: Co-localization analysis of image microscopy
package_version: 0.1.0
authors:
  - name: Mahmoud Ahmed

>     url: https://github.com/MahShaaban

Reviewer: ```suggestion
    url: https://github.com/MahShaaban
    twitter: Ma7moudsha3ban
```

Reviewer: If you want, making this change to YAML would provide a link to your twitter profile via twitter icon as in this example: https://ropensci.org/blog/2018/11/13/antarctic/

date: 2018-11-11
categories: blog
topicid:
tags:
- r

> - package

Reviewer: ```suggestion
- package
- review
```

- onboarding
- imaging
- reproducibility
---

> A few months ago, I wasn't sure what to expect when reading microscopy images. I looked at the accompanying graph to understand the data or the point the authors is trying to make. Often, the graph represents one or more measures of the so-called co-localization, but I couldn’t figure out how to interpret them. It turned out; reading the images is simple. Cells are simultaneously stained by two dyes (say, red and green) for two different proteins. The color turns yellow in the merged image when the two proteins localize in proximity. 

Reviewer: authors is trying -> authors were trying

> A few months ago, I wasn't sure what to expect when reading microscopy images. I looked at the accompanying graph to understand the data or the point the authors is trying to make. Often, the graph represents one or more measures of the so-called co-localization, but I couldn’t figure out how to interpret them. It turned out; reading the images is simple. Cells are simultaneously stained by two dyes (say, red and green) for two different proteins. The color turns yellow in the merged image when the two proteins localize in proximity. 

Reviewer: First two sentences might be clarified by saying you're looking at images in a published paper

> A recent publication from our laboratory included microscopy images and quantification graph. The standard analysis protocol is to use [ImageJ](https://imagej.nih.gov/ij/index.html) with a specialized plug-in for the co-localization analysis such as [coloc2](https://imagej.net/Coloc_2) or [Fiji](https://fiji.sc). The images were processed one at a time; subtracting the background, selecting regions of interest and calculating the co-localization statistics. Luckily for me, I didn’t do the work myself, but it took a good chunk of my colleague precious time. Even worse, she had to repeat this whole process more than once with minor changes in the parameters each time. At the end, there was no way to make this process reproducible, and we included only the final output in an otherwise fully reproducible article.

Reviewer: and quantification graph -> and a quantification graph

> A recent publication from our laboratory included microscopy images and quantification graph. The standard analysis protocol is to use [ImageJ](https://imagej.nih.gov/ij/index.html) with a specialized plug-in for the co-localization analysis such as [coloc2](https://imagej.net/Coloc_2) or [Fiji](https://fiji.sc). The images were processed one at a time; subtracting the background, selecting regions of interest and calculating the co-localization statistics. Luckily for me, I didn’t do the work myself, but it took a good chunk of my colleague precious time. Even worse, she had to repeat this whole process more than once with minor changes in the parameters each time. At the end, there was no way to make this process reproducible, and we included only the final output in an otherwise fully reproducible article.

Reviewer: You might omit some of this "Luckily for me, I didn’t do the work myself, but it took a good chunk of my colleague precious time. Even worse, she had to repeat this whole process more than once with minor changes in the parameters each time. " and condense by saying the work was done manually and repeated with minor changes in the parameters each time.

> For those two specific reasons; the need for manual work and the lack of reproducibility, I thought it would be great if I could do this analysis in [R](https://cran.r-project.org). There are several [R](https://cran.r-project.org) packages for image processing such as `EBImage`, `imager` and `magick`. These packages provided the functionality needed for conducting this analysis. So I put the pieces together in an [R](https://cran.r-project.org) package to simplify this analysis for my future self and for others who might find it useful. This is `colocr`, a simple [R](https://cran.r-project.org) package for conducting co-localization analysis of fluorescence microscopy images. In particular, it semi-automates the selections of regions of interest-arguably the most time-consuming step- and comes with a shiny app for visually comparing the outputs of different parameters before writing a final script. 

Reviewer: Please omit links to R

> ## How would've we done it differently?

Reviewer: ```suggestion
## How would we have done it differently?
```

> In this recent article I referred to earlier, we assessed the co-localization of several proteins with RKIP (Raf-kinas inhibitor protein) in DU145 prostate cancer cell line. For each of these proteins, we generated multiple frames, a hundred in total. We then processed the images one by one and repeated the process with each minor change in the parameters. Now we can use `colocr`, in just a few lines of code, to apply the same analysis.

Reviewer: Can you provide a link to the article?

A selection of the images were presented in the article in a similar figure.

> ![](/img/blog-images/2018-11-11-colocr/images.png)

Reviewer: should there be a heading above the red dye column?

Author: The labels for the red dye images are displayed on the left.

> We can calculate for the [**Pearson's** Correlation Coefficients](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient) (PCC) of multiple images at once using the following chunk of code.

Reviewer: ```suggestion
We can calculate the Pearson's Correlation Coefficients (PCC) of multiple images at once using the following chunk of code.
```

```r
## load required libraries
library(tidyverse)
library(colocr)

# load images
fls <- list.files('images', pattern = '*.jpg', recursive = TRUE)

# apply co-localization test
tst <- image_load(fls) %>%
  roi_select(threshold = 90, n = 3) %>%
  roi_test()
```

> And the results can be visualized in a graph.

Reviewer: Can you provide some text explanation of the plot? 
I'm having a hard time figuring out exactly what the correlations are and why for each marker labelled on the x-axis there is some spread along the x-axis.

Reviewer: Might help the reader's interest if you note what the plot allowed you to conclude about the co-localization.

Author: Added text explanation and conclusion.
The spread on the x-axis is just jitter produced by `ggplot2::geom_jitter()`. Otherwise, all data points would line and overlap.

![](/img/blog-images/2018-11-11-colocr/graph.png)

The following code is to generate the graph.

```r
## load required libraries
library(tidyverse)

# make the graph of calculated correlations
names(tst) <- str_split(fls, pattern = '/', simplify = TRUE)[, 1]

bind_rows(tst, .id = 'protein') %>%
  group_by(protein) %>%
  mutate(ave = mean(pcc),
         sd = sd(pcc),
         upper = ave + sd,
         lower = ave - sd) %>%
  ggplot(aes(x = protein, y = pcc)) +
  geom_jitter() +
  geom_point(aes(y = ave), color = 'red') +
  geom_linerange(aes(ymax = upper, ymin = lower), color = 'red') +
  lims(y = c(0, 1.1)) +
  labs(x = '', y = "Pearson's Correlation Coefficient (PCC)") +
  theme_bw()
```

> While developing `colocr`, I focused on breaking down the steps of the analysis into intuitive pieces. Hopefully, in a way that the potential user, or "myself in a few months", can understand easily. By keeping the number of the functions limited, the inputs and the outputs consistent and making things visual as much as possible, I hope using the package needs less investment by the user. I hope biologist who are not familiar with [R](https://cran.r-project.org) can still find this work useful. One thing helped me organize and iterate on the code. This was developing a shiny app as an interface to the package.

Reviewer: biologist -> biologists

> ## Developing a shiny app in an R package

Reviewer: capitalize Shiny each time I think

At first, I thought of adding a shiny app as an extra feature for users who might need it. I quickly realized that for this use case, selecting regions of interest might require trial-and-error on the side of the user to arrive at suitable parameters for the selection. Having to develop the app forced me to think about the organizing the functions, their input and output in a certain way. Namely, mapping the functions to the steps of the analysis that the user has to think about and making visual output a priority. Finally, testing the app exposed errors in the intermediates which might have been buried in an otherwise seemingly fine [R](https://cran.r-project.org) objects. Overall, I learned a thing or two about writing shiny apps to go with [R](https://cran.r-project.org) packages when possible even if it won't make the cut in the release.

* I found it useful to work on the app as early as possible and improving it side by side with the core functionality. As explained earlier, this helped me to think about the inputs and the output from a user's perspective.

> *  I learnt to isolate the source code for the app from that of the package functions as much as possible. One may host the app code in a separate repository and including it in the project as a **git submodule**. Some might find this necessary, but I found it easier this way to keep tracking of the changes in the app and testing it on Travis.

Reviewer: keep tracking -> keep track

> *  I learnt to isolate the source code for the app from that of the package functions as much as possible. One may host the app code in a separate repository and including it in the project as a **git submodule**. Some might find this necessary, but I found it easier this way to keep tracking of the changes in the app and testing it on Travis.

Reviewer: omit boldface of git submodule

> ## Submitting R packages for reviewing on ROpenSci

Reviewer: ROpenSci -> rOpenSci 
Please use this capitalization throughout post. No need to include URL since it's on our site

> `colocr` was my second submission on [ROpenSci](https://ropensci.org). In both cases, I had positive experiences. So I would like to thank the [ROpenSci](https://ropensci.org) editor, the reviewers ([@haozhu233](https://github.com/haozhu233), [@seaaan](https://github.com/seaaan)) and [@jeroen](https://github.com/jeroen) who helped us to resolve issues related to running the package on windows. At a risk of an understatement, here is three benefits I got from submitting my packages to [ROpenSci](https://ropensci.org) which others might find encouraging doing the same.

Reviewer: ```suggestion
`colocr` was my second submission to [ROpenSci](https://ropensci.org). In both cases, I had positive experiences. So I would like to thank the [ROpenSci](https://ropensci.org) editor, the reviewers ([Hao Zhu](https://github.com/haozhu233), [Sean Hughes](https://github.com/seaaan)) and [Jeroen Ooms](https://github.com/jeroen) who helped us to resolve issues related to running the package on Windows. At a risk of an understatement, here are three benefits I got from submitting my packages to [ROpenSci](https://ropensci.org) which others might find encouraging doing the same.
```

Reviewer: suggested changes above to spell out people's names, and made minor grammar corrections

* Reviewing the source could at an early stage help to spot mistakes and introducing improvement to the code that might escape the authors themselves.
* Often the editor and the reviewers are experts in [R](https://cran.r-project.org) and the subject matter, so they can offer ideas and suggestion to improve the design of the package. Adding to being an early test users who provided useful insights.
* Writing and distributing code might not be everyone's strongest skill. And certain aspects of it might be challenging. The [ROpenSci](https://ropensci.org) team provide support during the process I found encouraging.

> ## Things to improve

Reviewer: It's great that you've identified things to improve. Perhaps you could open them as issues in the repo and link to them - might draw people to the repo who might be interested to help.
